File: crates/core/src/progress.rs
use crate::conversation::ConversationProvider;
use crate::ids::{FrameId, NodePath, RunId, StepId, WorkflowId};
use crate::run::event::{
    BranchSelection, LoopIterationOutcome, NodeStatus, RunReason, RunStatus, StreamKind,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ProviderEvent {
    ToolUse { provider: ConversationProvider, tool: String, detail: Option<String> },
    Status { provider: ConversationProvider, message: String },
    Error { provider: ConversationProvider, message: String },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LiveEvent {
    RunStarted {
        run_id: RunId,
        workflow_id: WorkflowId,
        node_count: usize,
    },
    NodeStarted {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        node_kind: String,
        provider: Option<ConversationProvider>,
        attempt: u32,
    },
    NodeSkipped {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        reason: String,
    },
    BranchSelected {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        case_index: usize,
        selection: BranchSelection,
    },
    LoopIterationStarted {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        iteration: u32,
        max_iterations: u32,
    },
    LoopIterationFinished {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        iteration: u32,
        max_iterations: u32,
        outcome: LoopIterationOutcome,
    },
    StepOutput {
        stream: StreamKind,
        chunk: String,
    },
    ProviderToolUse {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        provider: ConversationProvider,
        tool: String,
        detail: Option<String>,
    },
    ProviderStatus {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        provider: ConversationProvider,
        message: String,
    },
    ProviderError {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        provider: ConversationProvider,
        message: String,
    },
    NodeFinished {
        frame_id: FrameId,
        node_path: NodePath,
        user_id: Option<StepId>,
        status: NodeStatus,
        exit_code: Option<i32>,
        duration_ms: Option<u128>,
        stdout_path: Option<String>,
        stderr_path: Option<String>,
    },
    RunFinished {
        status: RunStatus,
        reason: RunReason,
    },
}

pub trait ProgressSink {
    fn is_enabled(&self) -> bool {
        false
    }

    fn emit(&mut self, _event: LiveEvent) {}
}

pub trait StepProgressSink {
    fn is_enabled(&self) -> bool {
        false
    }

    fn step_output(&mut self, _stream: StreamKind, _chunk: &str) {}

    fn provider_event(&mut self, _event: ProviderEvent) {}
}

#[derive(Debug, Default)]
pub struct NoopProgressSink;

impl ProgressSink for NoopProgressSink {}
impl StepProgressSink for NoopProgressSink {}

File: packages/rigg/src/cli/progress.ts
import type { RunProgressEvent } from "../run/progress"

export class TerminalProgressSink {
  readonly #stream: NodeJS.WriteStream

  constructor(stream: NodeJS.WriteStream) {
    this.#stream = stream
  }

  emit(event: RunProgressEvent): void {
    switch (event.kind) {
      case "run_started":
        this.#stream.write(`[run] id=${event.run_id} workflow=${event.workflow_id} nodes=${event.node_count}\n`)
        return
      case "node_started":
        this.#stream.write(
          `[start] node=${event.user_id ?? event.node_path} kind=${event.node_kind} provider=${event.provider ?? "-"} attempt=${event.attempt}\n`,
        )
        return
      case "node_skipped":
        this.#stream.write(`[skip] node=${event.user_id ?? event.node_path} reason=${event.reason}\n`)
        return
      case "branch_selected":
        this.#stream.write(
          `[branch] node=${event.user_id ?? event.node_path} case=${event.case_index} kind=${event.selection}\n`,
        )
        return
      case "loop_iteration_started":
        this.#stream.write(
          `[loop] node=${event.user_id ?? event.node_path} iteration=${event.iteration}/${event.max_iterations} status=started\n`,
        )
        return
      case "loop_iteration_finished":
        this.#stream.write(
          `[loop] node=${event.user_id ?? event.node_path} iteration=${event.iteration}/${event.max_iterations} status=${event.outcome}\n`,
        )
        return
      case "step_output":
        this.#stream.write(event.chunk)
        return
      case "provider_tool_use":
        this.#stream.write(
          `[tool] node=${event.user_id ?? event.node_path} provider=${event.provider} tool=${event.tool}${event.detail === null ? "" : ` ${event.detail}`}\n`,
        )
        return
      case "provider_status":
        this.#stream.write(
          `[progress] node=${event.user_id ?? event.node_path} provider=${event.provider} ${event.message}\n`,
        )
        return
      case "provider_error":
        this.#stream.write(
          `[error] node=${event.user_id ?? event.node_path} provider=${event.provider} ${event.message}\n`,
        )
        return
      case "node_finished":
        this.#stream.write(
          `[done] node=${event.user_id ?? event.node_path} status=${event.status} exit=${String(event.exit_code)} duration=${formatDuration(event.duration_ms)} stdout=${event.stdout_path ?? "-"} stderr=${event.stderr_path ?? "-"}\n`,
        )
        return
      case "run_finished":
        this.#stream.write(`[run] status=${event.status} reason=${event.reason}\n`)
    }
  }
}

function formatDuration(durationMs: number | null): string {
  return durationMs === null ? "-" : `${durationMs}ms`
}
